src/BuildScriptGenerator/Java/JavaPlatform.cs

```diff
@@ -0,0 +1,417 @@
+﻿// --------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// --------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Oryx.BuildScriptGenerator.Exceptions;
+using Microsoft.Oryx.Detector;
+using Microsoft.Oryx.Detector.Java;
+
+namespace Microsoft.Oryx.BuildScriptGenerator.Java
+{
+    /// <summary>
+    /// Java Platform.
+    /// </summary>
+    internal class JavaPlatform : IProgrammingPlatform
+    {
+        /// <summary>
+        /// The tar-gz option for Java modules.
+        /// </summary>
+        private readonly BuildScriptGeneratorOptions _commonOptions;
+        private readonly JavaScriptGeneratorOptions _javaScriptGeneratorOptions;
+        private readonly IJavaVersionProvider _javaVersionProvider;
+        private readonly IMavenVersionProvider _mavenVersionProvider;
```

Comment:
nit: java can be built using gradle or ant etc as well. Should we consider them here as well? so instead of a IMavenVersionProvider and MavenInstaller should we use a wrapper IJavaBuilderVersionProvider and JavaBuildToolInstaller?

Reply:
Regarding supporting different build systems: I did not add Ant as it did not look like a popular way of doing things currently and many seem to be using Maven. Regarding Gradle, yeah I think we should add support for it but seems like it is mostly used for Android based apps (which is also a valid scenario for codespaces since your could have a github repo for that those kind of apps and want to 'develop' in codespaces environment). I can create a separate PR for adding support for it though.

Regarding wrapper, I do not feel too strong about it. So far we have been having separate installers and providers for our tools and we could do the same for Gradle too. what do you think?